src/engine.rs

```diff
@@ -0,0 +1,187 @@
+use crate::net::http::request::Request;
+use crate::net::http::response::Response;
+use crate::timing::{Timing, TimingTable};
+use crate::types::ParseError;
+use cookie::CookieJar;
+use core::fmt::Debug;
+use core::str::FromStr;
+use reqwest::header::HeaderName;
+use reqwest::Method;
+use url::Url;
+
+/// Response that is returned from the fetch function
+pub struct FetchResponse {
+    /// Request that has been send
+    pub request: Request,
+    /// Response that has been received
+    pub response: Response,
+    /// Document tree that is made from the response
+    pub document: DocumentHandle,
+    /// Parse errors that occurred while parsing the document tree
+    pub parse_errors: Vec<ParseError>,
+    /// Rendertree that is generated from the document tree and css tree
+    pub render_tree: String,
+    /// Timing table that contains all the timings
+    pub timings: TimingTable,
+}
+
+impl Debug for FetchResponse {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        writeln!(f, "Request:")?;
+        writeln!(f, "{}", self.request)?;
+        writeln!(f, "Response:")?;
+        writeln!(f, "{}", self.response)?;
+        writeln!(f, "Document tree:")?;
+        writeln!(f, "{}", self.document)?;
+        writeln!(f, "Parse errors:")?;
+        for error in &self.parse_errors {
+            writeln!(f, "  ({}:{}) {}", error.line, error.col, error.message)?;
+        }
+        writeln!(f, "Render tree:")?;
+        writeln!(f, "{}", self.render_tree)?;
+        writeln!(f, "Timings:")?;
+        writeln!(f, "{}", self.timings)?;
+
+        Ok(())
+    }
+}
+
+fn fetch_url(
+    method: &str,
+    url: &str,
+    headers: Headers,
+    cookies: CookieJar,
+) -> Result<FetchResponse, Error> {
+    let mut http_req = Request::new(method, url, "HTTP/1.1");
+    http_req.headers = headers.clone();
+    http_req.cookies = cookies.clone();
+
+    let parts = Url::parse(url);
+    if parts.is_err() {
+        return Err(Error::Generic(format!("Failed to parse URL: {}", url)));
+    }
+
+    let mut fetch_response = FetchResponse {
+        request: http_req,
+        response: Response::new(),
+        document: DocumentBuilder::new_document(),
+        parse_errors: vec![],
+        render_tree: String::new(),
+        timings: TimingTable::default(),
+    };
+
+    // For now, we do a DNS lookup here. We don't use this information yet, but it allows us to
+    // measure the DNS lookup time.
+    fetch_response.timings.start(Timing::DnsLookup);
+
+    let mut resolver = crate::dns::Dns::new();
+    let res = resolver.resolve(parts.unwrap().host_str().unwrap(), ResolveType::Ipv4);
+    if res.is_err() {
+        return Err(Error::Generic(format!("Failed to resolve domain: {}", url)));
+    }
+
+    fetch_response.timings.end(Timing::DnsLookup);
+
+    // Fetch the HTML document from the site
+    fetch_response.timings.start(Timing::ContentTransfer);
+
+    let m = Method::from_str(method).unwrap();
```

Comment:
Mostly converting our http request/response to reqwest objects. This can be done better, but this will be removed later on anyway